src/esp/physics/PhysicsManager.cpp

```diff
@@ -0,0 +1,481 @@
+  LOG(INFO) << "Initializing Base Physics Engine...";
+
+  physicsNode_ = node;
+  //! Create new scene node
+  sceneNode_ = std::make_shared<physics::RigidObject>(physicsNode_);
+  initialized_ = true;
+
+  return true;
+}
+
+PhysicsManager::~PhysicsManager() {
+  LOG(INFO) << "Deconstructing PhysicsManager";
+}
+
+bool PhysicsManager::addScene(
+    const assets::AssetInfo& info,
+    assets::PhysicsSceneAttributes& physicsSceneAttributes,
+    std::vector<assets::CollisionMeshData>& meshGroup) {
+  // Test Mesh primitive is valid
+  for (assets::CollisionMeshData& meshData : meshGroup) {
+    if (!isMeshPrimitiveValid(meshData)) {
+      return false;
+    }
+  }
+
+  switch (info.type) {
+    case assets::AssetType::INSTANCE_MESH:
+      break;  // ._semantic.ply mesh data
+    case assets::AssetType::FRL_INSTANCE_MESH:
+      break;  // FRL mesh
+    default:
+      break;  // GLB mesh data
+  }
+
+  //! Initialize scene
+  bool sceneSuccess =
+      sceneNode_->initializeScene(physicsSceneAttributes, meshGroup);
+
+  return sceneSuccess;
+}
+
+int PhysicsManager::addObject(const int resObjectID, DrawableGroup* drawables) {
+  const std::string configFile = resourceManager_->getObjectConfig(resObjectID);
+
+  //! Test Mesh primitive is valid
+  std::vector<assets::CollisionMeshData> meshGroup =
+      resourceManager_->getCollisionMesh(configFile);
+  assets::PhysicsObjectAttributes physicsObjectAttributes =
+      resourceManager_->getPhysicsObjectAttributes(configFile);
+  for (assets::CollisionMeshData& meshData : meshGroup) {
+    if (!isMeshPrimitiveValid(meshData)) {
+      return false;
+    }
+  }
+
+  //! Instantiate with mesh pointer
+  int nextObjectID_ = makeRigidObject(meshGroup, physicsObjectAttributes);
+  if (nextObjectID_ < 0) {
+    LOG(ERROR) << "Initialize unsuccessful";
+    return -1;
+  }
+
+  //! Draw object via resource manager
+  //! Render node as child of physics node
+  resourceManager_->loadObject(
+      configFile, existingObjects_.at(nextObjectID_).get(), drawables);
+
+  return nextObjectID_;
+}
+
+int PhysicsManager::addObject(const std::string configFile,
+                              DrawableGroup* drawables) {
+  int resObjectID = resourceManager_->getObjectID(configFile);
+  //! Invoke resourceManager to draw object
+  int physObjectID = addObject(resObjectID, drawables);
+  return physObjectID;
+}
+
+int PhysicsManager::removeObject(const int physObjectID) {
+  if (existingObjects_.count(physObjectID) == 0) {
+    LOG(ERROR) << "Failed to remove object: no object with ID " << physObjectID;
+    return -1;
+  }
+  existingObjects_.at(physObjectID)->removeObject();
+  existingObjects_.erase(physObjectID);
+  deallocateObjectID(physObjectID);
+  return physObjectID;
+}
+
+int PhysicsManager::allocateObjectID() {
+  if (!recycledObjectIDs_.empty()) {
+    int recycledID = recycledObjectIDs_.back();
+    recycledObjectIDs_.pop_back();
+    return recycledID;
```

Comment:
Just a note -- on the GPU side of things, object ID is currently just 32 bit, in case of per-primitive object IDs just 22 bit (due to them being in a float texture), and I was thinking about reducing it to just 16 to improve performance. So maybe we should add some assertions to the shaders to avoid th IDs getting truncated.